### PR TITLE
Refactor HoC events queries to use MySQL instead of Solr

### PR DESCRIPTION
**Pull request by wjordan**
Depends on ~#22644~ [merged].

This PR updates the https://hourofcode.com/events/all and related pages, and the https://code.org/volunteer/local and related pages to use MySQL queries instead of Solr.

Here are sample execution plans and run-times on our production dataset for these queries:

`Forms.events_by_country`:
```
$ bundle exec ruby -r./deployment -r./pegasus/helper_modules/forms -e "puts Forms.events_by_country('HocSignup2017', explain: true)"
SELECT `location_country_code_s` AS `country_code`, count(*) AS `count` FROM `forms` WHERE ((`kind` = 'HocSignup2017') AND (`location_country_code_s` != 'US')) GROUP BY `location_country_code_s`
+----+-------------+-------+------------+-------+---------------------------------------------------------------------------+-------------------------+---------+------+--------+----------+--------------------------+
| id | select_type | table | partitions | type  | possible_keys                                                             | key                     | key_len | ref  | rows   | filtered | Extra                    |
+----+-------------+-------+------------+-------+---------------------------------------------------------------------------+-------------------------+---------+------+--------+----------+--------------------------+
|  1 | SIMPLE      | forms | NULL       | range | kind,forms_kind_index,forms_kind_email_name_index,index_kind_country_code | index_kind_country_code | 159     | NULL | 116038 |   100.00 | Using where; Using index |
+----+-------------+-------+------------+-------+---------------------------------------------------------------------------+-------------------------+---------+------+--------+----------+--------------------------+

[...]
198 rows in set (0.06 sec)
```

`Forms.events_by_state`:
```
$ bundle exec ruby -r./deployment -r./pegasus/helper_modules/forms -e "puts Forms.events_by_state('HocSignup2017', explain: true)"
SELECT processed_data->>'$.location_state_code_s' AS `state_code`, count(*) AS `count` FROM `forms` WHERE ((`kind` = 'HocSignup2017') AND (`location_country_code_s` = 'US')) GROUP BY processed_data->>'$.location_state_code_s'
+----+-------------+-------+------------+------+---------------------------------------------------------------------------+-------------------------+---------+-------------+--------+----------+----------------------------------------------+
| id | select_type | table | partitions | type | possible_keys                                                             | key                     | key_len | ref         | rows   | filtered | Extra                                        |
+----+-------------+-------+------------+------+---------------------------------------------------------------------------+-------------------------+---------+-------------+--------+----------+----------------------------------------------+
|  1 | SIMPLE      | forms | NULL       | ref  | kind,forms_kind_index,forms_kind_email_name_index,index_kind_country_code | index_kind_country_code | 159     | const,const | 112602 |   100.00 | Using where; Using temporary; Using filesort |
+----+-------------+-------+------------+------+---------------------------------------------------------------------------+-------------------------+---------+-------------+--------+----------+----------------------------------------------+

[...]
53 rows in set (2.31 sec)
```

`Forms.events_by_name`:
```
$ bundle exec ruby -r./deployment -r./pegasus/helper_modules/forms -e "puts Forms.events_by_name('HocSignup2017', 'US', 'CA', explain: true)"
SELECT DISTINCT data->>'$.organization_name_s' AS `name`, processed_data->>'$.location_city_s' AS `city` FROM `forms` WHERE ((`kind` = 'HocSignup2017') AND (`location_country_code_s` = 'US') AND (processed_data->>'$.location_state_code_s' = 'CA')) ORDER BY `city`, `name`
+----+-------------+-------+------------+------+---------------------------------------------------------------------------+-------------------------+---------+-------------+--------+----------+----------------------------------------------+
| id | select_type | table | partitions | type | possible_keys                                                             | key                     | key_len | ref         | rows   | filtered | Extra                                        |
+----+-------------+-------+------------+------+---------------------------------------------------------------------------+-------------------------+---------+-------------+--------+----------+----------------------------------------------+
|  1 | SIMPLE      | forms | NULL       | ref  | kind,forms_kind_index,forms_kind_email_name_index,index_kind_country_code | index_kind_country_code | 159     | const,const | 112602 |   100.00 | Using where; Using temporary; Using filesort |
+----+-------------+-------+------------+------+---------------------------------------------------------------------------+-------------------------+---------+-------------+--------+----------+----------------------------------------------+

[...]
1237 rows in set (0.41 sec)
```

### CODE REVIEW

**Comment by wjordan** [comment]
@breville here is the latest WIP for refactoring Solr queries to the MySQL equivalents. Beyond the existing minimal test coverage and some extra query tests I've added, I'm finding it hard to test these old legacy routes extensively against existing behavior, since the latter isn't very clearly defined anywhere, depends mostly on existing production data, etc. Hope you (or whoever is owning/maintaining these features moving forward) can more or less take over this work from here.

**Comment by breville** [comment]
Thanks @wjordan.  I have some upcoming work on Pegasus forms and can start looking at it as part of that work.  Is there a preferred timeline for this work that means it should get prioritized?

**Comment by wjordan** [comment]
@breville for timeline/deadline priority I'd check with @poorvasingal and @jeremydstone as stakeholders. If I recall, Jeremy is concerned about long-term operational cost/effort of maintaining our self-hosted Solr instance, and Poorva is concerned about user-data removal/purge features not currently implemented in Solr-indexed collections. I don't know if either has any important deadlines affecting priority but they could probably clarify.

**Comment by joshlory** [comment]
Are there parts of this PR that can be front-loaded / merged ahead of this PR to reduce its diff?

I think ideally we'd add the new code path so it could live next to the old SOLR code.  Then after testing flipping the switch to MySQL 5.7 (maybe using a DCDO flag?) we could verify the new behavior & deprecate the old code.

**Comment by poorvasingal** [comment]
In terms of timeline, this seemed like an "as soon as we can" sort of thing. There are privacy implications here, so the earlier we can get it cleaned up, the better. But I think it's fine for this to slip by a little.

**Comment by breville** [comment]
Thank you @wjordan for such a big change!  

I built on it by working through all of the affected Pegasus forms and now have them working with this change, and fixed some other small forms issues along the way.  There were quite a few commits in the journey, with notable moments including ordering latitude & longitude, converting the hoc events form to also use MySQL, and making sure the additional processed fields in a class submission were written to MySQL instead of just Solr.

Right now, it appears that we've taken Solr out of the Pegasus forms system altogether, with it no longer required for processing forms or consuming their submitted contents, either.  

@tanyaparker has helped test all of the forms and their related pages manually on an adhoc instance.  

A "one off" backfill script has been prepped in https://github.com/code-dot-org/code-dot-org/pull/24224 to store some processed data, that was previously stored only in Solr for some form submissions, in MySQL.

File: pegasus/forms/class_submission.rb [hunk 1]
```diff
@@ -145,6 +147,9 @@ def self.process_with_ip(data, created_ip)
     {}.tap do |results|
       location = search_for_address(data['school_address_s'])
       results.merge! location.to_solr if location
```
Comment (aoby) [inline pegasus/forms/class_submission.rb]: Should this still be called `to_solr`? How about something more generic like `process` or `summarize` (or `to_summary` if you prefer?

File: pegasus/forms/class_submission.rb [hunk 2]
```diff
@@ -172,62 +177,88 @@ def self.additional_data(data)
...
-        data['class_format_subcategory_s'] = class_format.sub(/^#{prefix}_/, '')
-      end
-    end
-
-    data['class_languages_all_ss'] = data['class_languages_ss'] - ['Other']
-    data['class_languages_all_ss'].concat(data['class_languages_other_ss'] || []).sort.uniq
-
-    # Create a case-insensitive version of the name for sorting.
-    data['school_name_sort_s'] = data['school_name_s'].downcase
+  def self.solr_query(params)
```
Comment (aoby) [inline pegasus/forms/class_submission.rb]: Same with this. Do we want to keep referring to solr after we're no longer using it? Why not just `query`?

File: pegasus/sites.v3/hourofcode.com/public/events/2013/index.haml [hunk 3]
```diff
@@ -3,38 +3,25 @@ title: <%= hoc_s(:events_all_title) %>
 layout: wide
 nav: events_nav
 ---
--SOLR = Solr::Server.new(host:CDO.solr_server)
+:ruby
+  require pegasus_dir 'helper_modules/forms'
+  hoc_year = 2013
+  kind = 'CSEdWeekEvent2013'
```
Comment (aoby) [inline pegasus/sites.v3/hourofcode.com/public/events/2013/index.haml]: nit: should these be constants?

Same with similar static values in other forms below.

**Comment by breville** [comment]
Our testing has focused on these three forms, with relevant URLs as shown:

#### ClassSubmission:
- create: code.org/schools/new
- approve: code.org/private/review-class
- view: code.org/learn/local

#### VolunteerEngineerSubmission2015
- create: code.org/volunteer
- view: code.org/volunteer/local

#### HocSignup2017:
- create: hourofcode.com/
- view: hourofcode.com/events (and sub pages)

**Comment by aoby** [comment]
We still have an open form with 2015 in the title? 🤕

**Pull request merged**